### PR TITLE
Add watch RPC

### PR DESCRIPTION
**Pull request by rleungx**

### CODE REVIEW

**Review by lhy1024 — APPROVED** [review]

File: proto/pdpb.proto [hunk 1]
```diff
@@ -113,6 +119,53 @@ message WatchGlobalConfigResponse {
...
+message GetRequest {
+    bytes key = 1;
+    bytes range_end = 2;
+    int64 limit = 3;
+    int64 revision = 4;
+}
+
+// copied part of https://github.com/etcd-io/etcd/blob/7dfd29b0cc7ce25337276dce646ca2a65aa44b4d/api/etcdserverpb/rpc.proto
+message GetResponse {
+    ResponseHeader header = 1;
```
Comment (YuJuncen) [inline proto/pdpb.proto:147]: I think the response header should be a trimmed of this:
https://github.com/etcd-io/etcd/blob/3cf2e79e16a7a0e08de0ac168c829b4df20ae644/api/etcdserverpb/rpc.proto#L391-L406

(At least I think we need to keep the `revision` field of this...)

Comment (rleungx) [reply proto/pdpb.proto:147]: Done. Add the revision in the corresponding response.

File: proto/pdpb.proto [hunk 2]
```diff
@@ -103,8 +103,9 @@ service PD {
 }
 
 message WatchGlobalConfigRequest {
-    string config_path = 1;
-    int64 revision = 2;
+    string source = 1;
```
Comment (glorv) [inline proto/pdpb.proto:106]: this is change will break the backward compatibility

Comment (rleungx) [reply proto/pdpb.proto:106]: It is changed by accident.

Comment (rleungx) [reply proto/pdpb.proto]: fixed

**Comment by rleungx** [comment]
BTW, should we name it storagepb instead of etcdpb? @binshi-bing @nolouch

**Comment by binshi-bing** [comment]
> BTW, should we name it storagepb instead of etcdpb? @binshi-bing @nolouch

+1 for storagepb or kvstoragepb

**Review by binshi-bing — APPROVED** [review]
LGTM

File: proto/storagepb.proto [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+// copied part of https://github.com/etcd-io/etcd/blob/7dfd29b0cc7ce25337276dce646ca2a65aa44b4d/api/etcdserverpb/rpc.proto
+message WatchResponse {
+    ResponseHeader header = 1;
+    int64 compact_revision = 2;
+    repeated Event events = 3;
+}
+
+// copied part of https://github.com/etcd-io/etcd/blob/7dfd29b0cc7ce25337276dce646ca2a65aa44b4d/api/etcdserverpb/rpc.proto
+message GetRequest {
```
Comment (HuSharp) [inline proto/storagepb.proto:68]: Do you think we need to add more comments to indicate that the `GetRequest` here corresponds to etcd's `RangeRequest`?

Comment (rleungx) [reply proto/storagepb.proto:68]: Both are ok if we are going to use another kind of storage, Get sounds more general.

**Comment by nolouch** [comment]
> > BTW, should we name it storagepb instead of etcdpb? @binshi-bing @nolouch
> 
> +1 for storagepb or kvstoragepb

I think storagepb or kvstoragepb maybe too general. we have TiKV and Tiflash storage.

**Comment by YuJuncen** [comment]
Perhaps we can delete the generated `etcdpb.pb.go` for now?
BTW, I'm wondering will changing the service name make implement the client harder?
Rest LGTM.

**Comment by rleungx** [comment]
> Perhaps we can delete the generated `etcdpb.pb.go` for now? BTW, I'm wondering will changing the service name make implement the client harder? Rest LGTM.

Yes, I noticed that and did a cleanup.

**Comment by rleungx** [comment]
> > > BTW, should we name it storagepb instead of etcdpb? @binshi-bing @nolouch
> > 
> > 
> > +1 for storagepb or kvstoragepb
> 
> I think `storagepb` or `kvstoragepb` maybe too general. we already have TiKV and TiFlash storage, it's confusing. I prefer `etcdpb` or `metastoragepb`.

change to `metastoragepb`

**Comment by rleungx** [comment]
@YuJuncen @nolouch comment addressed, PTAL

File: proto/storagepb.proto (hunk 3 above)
Comment (rleungx) [reply proto/storagepb.proto]: add a comment in RPC definition

**Review by YuJuncen — APPROVED** [review]
LGTM

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**